packages/custom-keybindings/src/node/custom-keybinding-server.ts

```diff
@@ -0,0 +1,126 @@
+        @inject(ILogger) protected readonly logger: ILogger,
+        @inject(KeybindingURI) protected readonly fileUri: URI
+    ) {
+
+        this.toDispose.push(watcherServer);
+
+        watcherServer.setClient({
+            onDidFilesChanged: p => this.onDidFilesChanged(p)
+        });
+
+        watcherServer.watchFileChanges(this.fileUri.toString()).then(id => {
+            this.toDispose.push(Disposable.create(() =>
+                watcherServer.unwatchFileChanges(id))
+            );
+        });
+
+        this.reconcileKeybindings();
+    }
+
+    dispose(): void {
+        this.toDispose.dispose();
+    }
+
+    protected onDidFilesChanged(params: DidFilesChangedParams): void {
+        if (this.areKeybindingsAffected(params.changes)) {
+            this.reconcileKeybindings();
+        }
+    }
+
+    /**
+     * Checks to see if the keybindings file was modified
+     * @param changes Changes
+     */
+    protected areKeybindingsAffected(changes: FileChange[]): boolean {
+        return changes.some(c => c.uri === this.fileUri.toString());
+    }
+
+    protected reconcileKeybindings(): void {
+        this.readKeybindings().then(keybindings =>
+            this.handleKeybindingChanges(keybindings)
+        );
+    }
+
+    protected readKeybindings(): Promise<JSON | undefined> {
+        return this.fileSystem.exists(this.fileUri.toString()).then(exists => {
+            if (!exists) {
+                return undefined;
+            }
+            return this.fileSystem.resolveContent(this.fileUri.toString()).then(({ stat, content }) => {
+                const strippedContent = jsoncparser.stripComments(content);
+                const errors: ParseError[] = [];
+                const keybindings = jsoncparser.parse(strippedContent, errors);
+                if (errors.length) {
+                    for (const error of errors) {
+                        this.logger.error("JSON parsing error", error);
+                    }
+                }
+
+                return Promise.resolve(keybindings);
+            });
+        }).catch(reason => {
+            if (reason) {
+                this.logger.error(`Failed to read keybindings ${this.fileUri}:`, reason);
+            }
+            return Promise.resolve(undefined);
+        });
+    }
+
+    protected handleKeybindingChanges(keybindings: any | undefined): void {
+
+        const rawBindings: RawKeybinding[] = [];
+
+        for (const keybinding of keybindings) {
+            rawBindings.push({
+                command: keybinding.command,
+                keybinding: keybinding.keybinding,
+                context: keybinding.context,
+                args: keybinding.args
+            });
+        }
+
+        this.fireEvent({ changes: rawBindings });
+        this.keybindings = rawBindings;
+    }
+
+    protected fireEvent(event: KeymapChangeEvent) {
+        this.logger.debug(log =>
+            log('onDidChangeKeymap:', event)
```

Comment:
Please create a child logger as in the terminal module otherwise we have no idea where this log comes from